stream-chat-android-state/src/main/java/io/getstream/chat/android/offline/plugin/listener/internal/SendMessageListenerImpl.kt

```diff
@@ -59,20 +68,29 @@ internal class SendMessageListenerImpl(
      * @param processedMessage [Message] returned from API response.
      * @return [Message] Updated message.
      */
-    private suspend fun handleSendMessageSuccess(channel: ChannelLogic, processedMessage: Message) {
+    private suspend fun handleSendMessageSuccess(
+        cid: String,
+        channel: ChannelLogic?,
+        thread: ThreadLogic?,
+        processedMessage: Message
+    ) {
+        if (channel == null && thread == null) {
+            return
+        }
         // Don't update latest message with this id if it is already synced.
         val latestUpdatedMessage = repos.selectMessage(processedMessage.id)
         if (latestUpdatedMessage?.syncStatus == SyncStatus.COMPLETED) {
             return
         }
-        processedMessage.enrichWithCid(channel.cid)
+        processedMessage.enrichWithCid(cid)
             .copy(
                 syncStatus = SyncStatus.COMPLETED,
                 syncDescription = null
             )
             .also {
                 repos.insertMessage(it)
-                channel.upsertMessage(it)
+                channel?.upsertMessage(it)
+                thread?.upsertMessage(it)
```

Comment:
This called my attention. Looks like we are always adding the message to the thread logic here. 

Shouldn't we check that the message belongs to a thread? 

If we add messages that don't belong to threads, we are going to use more memory than necessary, because we need the messages inside a map with more messages there, more memory we use. 

Let's make sure that we just add messages to threadLogic when they belong to a thread. 

This observation is also valid for other places.

Reply:
Use [`channelFromMessage`, `threadFromMessage` in LogicRegistry](https://github.com/GetStream/stream-chat-android/pull/3959/files#diff-03d341089e90e67508c7d008eba722373e032ce0ccd16d622b0c40e155640e45R111-R122) to know whether message is belong to channel or thread.